File: portal/components/manageManuscript/manageManuscript.go
package manageManuscript

import (
	"gitlab.bbinfra.net/3estack/alexandria/portal/util"
	"html/template"
)

var uploadTemplate = `
{{define "upload"}}
    <form enctype="multipart/form-data" class="uploadForm" id="uploadControl">
        <label class="uploadForm__label" for="uploadTrigger">Upload or verify manuscript:</label>
        <input class="uploadForm__input" type="file" name="file" id="uploadTrigger">
    </form>
{{end}}
`
var manageManuscriptTemplate = `
{{define "manageManuscript"}}
<div class="manageDocument">
  {{template "upload"}}
  <div class="notification" id="alert"></div>
  <script src="https://unpkg.com/axios/dist/axios.min.js"></script>
  <script src="{{.JsUrl}}"></script>
  <script>
    console.log("Going to initialize the context");
    var context = {
      subjectId: {{.SubjectId}},
      initialIsUploadNeeded: {{.InitialIsUploadNeeded}},
      updateUrlComponent: {{.UpdateUrlComponent}},
      downloadControlId: {{.DownloadControlId}},
    };
    console.log("Initialized the context");
    linkManageDocument(context);
  </script>
</div>
{{end}}
`

func ParseManageManuscriptTemplate(name string) *template.Template {
	return util.ParseTemplates(name,
		uploadTemplate, manageManuscriptTemplate)
}

type ManageManuscriptContext struct {
	SubjectId             string
	InitialIsUploadNeeded bool
	JsUrl                 string
	UpdateUrlComponent    string
	DownloadControlId     string
}

type PortalManuscriptResponse struct {
	Message      string
	UploadNeeded bool
	IsWarning    bool
}

File: portal/components/manageManuscript/manageManuscript.js
function linkManageDocument(context) {
    "use strict";
    console.log("Running linkManageDocument");
    var subjectId = context.subjectId;
    var uploadTriggerControl = document.querySelector("#uploadTrigger");
    var alertControl = document.querySelector('#alert');
    var downloadControlId = context.downloadControlId
    var downloadControl = document.querySelector("#" + downloadControlId)

    uploadTriggerControl.addEventListener("change", addFile);
    setUploadNeeded(context.initialIsUploadNeeded);

    function setUploadNeeded(uploadNeeded) {
        console.log("setUploadNeeded: " + uploadNeeded);
        if(uploadNeeded) {
            downloadControl.disabled = true
        } else {
            downloadControl.disabled = false
        }
        console.log("done setUploadNeeded")
    }

    function addFile(e) {
        console.log("Doing addFile")
        var file = e.target.files[0];
        if(!file){
            return
        }
        upload(file);
    }

    function upload(file) {
        var url = "/" + context.updateUrlComponent +  "/" + subjectId;
        console.log("Doing upload with url: " + url);
        var formData = new FormData();
        formData.append("file", file);
        post(url, formData)
            .then(onResponse)
            .catch(onResponse);
    }

    function post(url, data) {
        return axios.post(url, data)
            .then(function (response) {
                return response;
            }).catch(function (error) {
                return error.response;
            });
    }

    function onResponse(response) {
        if(response.status >= 400) {
            showResponse("error", response.data);
        } else {
            console.log("Got successful response: " + theResponse)
            var theResponse = response.data;
            var responseClass = "success";
            if(theResponse.IsWarning) {
                responseClass = "error"
            }
            showResponse(responseClass, theResponse.Message);
            setUploadNeeded(theResponse.UploadNeeded)
        }
    }

    function showResponse(responseClass, response) {
        alertControl.classList.remove("success");
        alertControl.classList.remove("error");
        alertControl.classList.add(responseClass);
        alertControl.innerHTML = response;
    }
}
